File: models/content.js
// @flow
import mongoose     from 'mongoose'
import sanitizeHtml from 'sanitize-html'
import {Parser,
	DomHandler}     from 'htmlparser2'
import {mergeWith,
	uniq,
	map,
	difference}     from 'lodash'
import Video        from './video.js'
import Image        from './image.js'
import DomUtils     from 'domutils'
import XXHash       from 'xxhash'

mongoose.Promise = global.Promise // Required to squash a deprecation warning

const schema = new mongoose.Schema({
	uri: {
		type:      String,
		unique:    true,
		minlength: 1,
		required:  true,
		set: function(uri) {
			// Track the previous URI so that when saved we can update all
			// links to the node with previous URI
			if (uri != this.uri)
				this._previousURI = this.uri
			
			return uri
		},
	},
	id: {
		type: String,
		unique: true,
		set: function(id) {
			this._idChanged = true
			return id
		}
	},
	order:       {type: Number, default: 0},
	body:        {type: String},
	description: {type: String},
	surtitle:    {type: String},
	redirect_uri:{type: String},
	hide:        {type: Boolean, default: false},
	further_reading_uri:
	             {type: String},
	title:       {type: String,  required:true},
	type:        {type: String,  default: "page"},
	has_sublist: {type: Boolean, default: false},
	authorship:  {
		type: String,
		get:  str =>   typeof str === 'string' ? str.split(/[:space:]*;[:space:]*/) : str,
		set:  input => input.join !== undefined ? input.join(';') : input,
	},
	contents:    {type: [{text: String, id: String}]},
}, {
	timestamps: true,
})

schema.index({
	body:        "text",
	title:       "text",
	description: "text",
}, {
	weights: {
		title:       10,
		description: 7,
		body:        4,
	},
})

schema.statics = {
	parentUriFragment(uri):string {
		return uri.split("/").slice(0, -1).join("/")
	},

	normalizeURI(uri):string {
		// Force the URI into acceptable format:
		return uri
			// All lowercase
			.toLowerCase()
			// Convert spaces and underscores to dashes (and multiple dashes)
			.replace(/[_ -]+/g, "-")
			// Remove any duplicate slashes
			.replace(/[\/]+/g, "/")
			// Remove any leading or trailing slashes or dashes
			.replace(/(^[\/-]+|[\/-]+$)/g, "")
			// Remove any remaining characters that don"t conform to the URL
			.replace(/[^a-z0-9-\/]+/g, "")
	},

	async getAllURIs():Promise<Array<string>> {
		const all_uris = await this.find().select('uri').exec()

		// Return just a list of the URI parts, prepended with the root URL
		return all_uris.map(uri => '/' + uri.uri)
	},

	getLinksInHTML(html):Array<string> {
		let links:Array<String> = []
		const parser = new Parser({
			onopentag(name, attribs) {
				if (name == "a" && attribs.href)
					// Track link without leading slash
					links.push(attribs.href)
			},
		})
		parser.write(html)
		parser.end()

		return uniq(links)
	},

	getImgSrcsInHTML(html):Array<string> {
		let images:Array<string> = []
		const parser = new Parser({
			onopentag(name, attribs) {
				if (name == "img" && attribs.src)
					// Strip leading slash
					images.push(attribs.src)
			},
		})
		parser.write(html)
		parser.end()

		return uniq(images)
	},

	getSanitizedHTML(html:string):string {
		// Force the body into an acceptable format
		// Allow only a super restricted set of tags and attributes
		let reduced_body = ""
		for(;;) {
			reduced_body = sanitizeHtml(html, {
				allowedTags: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'blockquote', 'p', 'a', 'ul', 'ol',
					'li', 'strong', 'em', 'table', 'thead', 'caption', 'tbody', 'tfoot', 'tr', 'th', 'td',
					'figure', 'abbr', 'img', 'aside', 'caption', 'cite', 'dd', 'dfn', 'dl', 'dt', 'figcaption',
					'sub', 'sup', 'i', 'br', 'hr', 'video', 'source'],
				allowedAttributes: mergeWith({
					th: ['colspan', 'rowspan'],
					td: ['colspan', 'rowspan'],
					a: ['class'],
					img: ['width', 'height'],
					video: ['src', 'controls'],
					source: ['src'],
				}, sanitizeHtml.defaults.allowedAttributes, (objValue, srcValue) => {
					if (Array.isArray(objValue) && Array.isArray(srcValue))
						return srcValue.concat(objValue)
				}),
				exclusiveFilter: frame => {
					// Remove certain empty tags
					return ["p", "a", "em", "strong"].includes(frame.tag) && !frame.text.trim() && !frame.children.length
				},
				textFilter: (text, stack) => {
					// Remove things not in a tag at all
					if (stack.length == 0)
						// If it"s not in a container class
						return text
						// Remove any non-whitespace characters
							.replace(/[^\s]+/g, "")
						// Remove any blank lines
							.replace(/[\n]+/g, "\n") // Unix
							.replace(/(\r\n)+/g, "\r\n") // Windows
					else
						return text
				},
			})

			if (reduced_body == html)
				break

			html = reduced_body
		}

		return html
	},

	getHTMLWithHeadingIDs(html:string) {
		let new_body
		let headings = []

		let handler = new DomHandler((err, dom) => {
			DomUtils.getElements({tag_name:"h1"}, dom, true).forEach(node => {
				let text = DomUtils.getText(node)
				let id = text
					// All lowercase
						.toLowerCase()
					// Make sure ampersands are covered
						.replace(/&amp/g, "and")
					// Convert spaces and underscores to dashes (and multiple dashes)
						.replace(/[_ -]+/g, "-")
					// Remove any duplicate slashes
						.replace(/[\/]+/g, "/")
					// Remove any leading or trailing slashes or dashes
						.replace(/(^[\/-]+|[\/-]+$)/g, "")
					// Remove any remaining characters that don"t conform to the URL
						.replace(/[^a-z0-9-\/]+/g, "")

				if (!node.attribs)
					node.attribs = {}

				node.attribs.id = id
				headings.push({text, id})
			})
			new_body = DomUtils.getOuterHTML(dom)
		})
		let parser = new Parser(handler)
		parser.write(html)
		parser.done()

		return {
			html: new_body,
			contents: headings,
		}
	},

	async getAllIdUriPairs():Promise<{uris: Array<string>, ids: Array<string>}> {
		const all_pages = await this.find().select('uri id').exec()

		let uris = []
		let ids  = []

		all_pages.forEach(page => {
			if (page.id && page.id !== "") {
				uris.push("/" + page.uri)
				ids.push("/" + page.id)
			}
		})

		return {ids, uris}
	},

	findWithNoAuthorship() {
		return this
			.find({ $or: [{authorship: {$exists: false}}, {authorship: ''}] })
			.where('type')
			.in(['level1', 'level2', 'level3', 'page'])
			.where('body')
			.ne('')
	},

	replaceHREFsWith(html:string, from:Array<string>, to:Array<string>):string {
		let handler = new DomHandler((err, dom) => {
			DomUtils.getElements({tag_name:"a"}, dom, true).forEach(node => {
				if (!node.attribs || !node.attribs.href)
					return

				for (let i = 0; i < from.length; i++) {
					if (node.attribs.href == from[i]) {
						node.attribs.href = to[i]
						return // don't bother with any other replacements
					}
				}
			})
			html = DomUtils.getOuterHTML(dom)
		})
		let parser = new Parser(handler)
		parser.write(html)
		parser.done()

		return html
	},

	getLineageFromURI(uri) {
		let fragments = []
		let parent = uri
		while (parent != "") {
			fragments.push(parent)
			parent = this.parentUriFragment(parent)
		}

		return fragments.reverse()
	},

	findBreadcrumbsForURI(uri) {
		return this.findFromURIs(
				this.getLineageFromURI(this.parentUriFragment(uri))
			)
			.select("title uri")
			.sort("uri")
	},

	findFromURIs(uris) {
		return this
			.find()
			.where("uri").in(uris.map(uri => uri.replace(/^\//, '')))
	},

	findFromParentURI(parent) {
		return this
			.find()
		// Only match URIs prefixed with the parent without any following slashes
			.where("uri", new RegExp(parent != "" ? `^${parent}/[^/]+$` : "^[^/]+$"))
	},

	findAllBelowURI(uri) {
		return this
			.find()
			.where("uri", new RegExp(`^${uri}/.+$`))
	},

	findFromAdjacentURI(uri) {
		return this.findFromParentURI(this.parentUriFragment(uri))
	},

	etagFromBuffer(buffer:Buffer|string, seed:number):string {
		if (typeof buffer !== 'Buffer')
			buffer = Buffer.from(buffer)

		return XXHash.hash64(
			buffer,
			// Convert to 32-bit int for hash
			seed >>> 0,
			// Hex output easily readable
			'hex'
		)
	},
}

schema
	.virtual("parent")
	.get(function() { return this.constructor.parentUriFragment(this.uri) })

schema
	.virtual("lineage")
	.get(function() { return this.constructor.getLineageFromURI(this.parent) } )

schema.methods = {
	getLinksInHTML() {
		return this.model('Content').getLinksInHTML(this.body)
	},

	getMatchedParagraph(regexp:RegExp) {
		let elements:Array<any> = []

		let handler = new DomHandler((err, dom) => {
			elements =
				DomUtils.getElements({tag_name:'p'}, dom, true).concat(
					DomUtils.getElements({tag_name:'li'}, dom, true),
					DomUtils.getElements({tag_name:'td'}, dom, true),
					DomUtils.getElements({tag_name:'h1'}, dom, true),
					DomUtils.getElements({tag_name:'h2'}, dom, true),
					DomUtils.getElements({tag_name:'h3'}, dom, true),
					DomUtils.getElements({tag_name:'h4'}, dom, true),
					DomUtils.getElements({tag_name:'h5'}, dom, true),
				)
		})
		let parser = new Parser(handler, {decodeEntities:true})
		parser.write(this.body)
		parser.done()

		for (let i = 0; i < elements.length; i++) {
			let match = DomUtils.getText(elements[i]).match(regexp)
			if (match != null)
				return match
		}

		return null
	},

	async getInvalidLinks() {
		const links = this.model("Content").getLinksInHTML(this.body)
			.filter(link => !link.match(/^([A-Za-z]+:)?\/\//)) // Filter external URLs
			.filter(link => !link.match(/\/feedback$/i)) // Filter feedback links
		const valid_pages = await this.model("Content").findFromURIs(links).select("uri").exec()
		const valid_videos = await Video.findFromURIs(links).select("uri").exec()
		const valid_images = await Image.findFromURIs(links).select("uri").exec()

		const valid_links = valid_pages.concat(valid_videos).concat(valid_images)// join them

		return difference(links, valid_links.map(content => '/' + content.uri))
	},

	// Retrieves an array of images used in this model
	getImages: function () {
		const images = this.model("Content").getImgSrcsInHTML(this.body)

		return this.model("Image")
			.findFromURIs(images)
			.select("_id uri")
			.exec()
	},

	getBreadcrumbs: function() {
		return this.constructor.findBreadcrumbsForURI(this.uri).exec()
	},

	async getChildren() {
		return this.model('Content').findFromParentURI(this.uri).exec()
	},

	getNextPage: async function() {
		// No next page for level 3
		if (this.type == "level3")
			return

		return (await this.model("Content")
			.find()
			/// The location can be below or adjacent to the current page
			.where("uri", new RegExp(`^(${this.uri}|${this.parent})/[^/]+$`))
			.select("-body")
			.where("title", this.title)
			// Only allow a type that could follow the current type
			.where("type").in(
				this.type == "level1" ? ["level2", "level3"] :
				this.type == "level2" ? ["level3"] :
				["level1", "level2", "level3"])
			// Make sure to favour the order of the types
			.sort("type")
			// Only need one of them
			.limit(1)
			.exec())[0]
	},

	setIDsForHeadings: function() {
		const heading_data = this.model("Content").getHTMLWithHeadingIDs(this.body)

		this.body     = heading_data.html
		this.contents = heading_data.contents
	},

	replaceHREFsWith: function(from:Array<String>, to:Array<String>) {
		this.body = this.model("Content").replaceHREFsWith(this.body, from, to)
	},
}

schema.pre("save", async function(next) {
	// Force the URIs into acceptable format:
	this.uri = this.model("Content").normalizeURI(this.uri)

	if (this.further_reading_uri)
		this.further_reading_uri = this.model("Content").normalizeURI(this.further_reading_uri)

	// Update any links to IDs in the body
	let {ids, uris} = await this.model("Content").getAllIdUriPairs()
	this.replaceHREFsWith(ids, uris)

	// Force the body into an acceptable format
	// Allow only a super restricted set of tags and attributes
	this.body = this.model("Content").getSanitizedHTML(this.body)

	// Generate table of contents from <h1> tags
	// And update IDs of header elements to match text
	this.setIDsForHeadings()

	// Save it
	next()
})

schema.post("save", async function(content) {

	// Need to replace any pages that use ID or old URI if there is one
	const has_new_uri = content._previousURI && content._previousURI != content.uri

	let find = []
	let replace = ['/' + content.uri, '/' + content.uri]

	if (content._idChanged && content.id !== "")
		find.push('/' + content.id)

	if (has_new_uri)
		find.push('/' + content._previousURI)

	if (find.length > 0) {
		// It was updated
		// Update links in text
		const matches = await content.model("Content")
			.find({ 
				body: { $regex : `(${find.join('|')})` },
				uri:  { $ne: content.uri },
			})
			.exec()

		matches.forEach(page => {
			page.replaceHREFsWith(find, replace)
			return page.save()
		})

		// And update child URIs that contain the parent URI
		const pages = await content.model("Content").findAllBelowURI(content._previousURI).exec()
		pages.forEach(async page => {
			page.uri = page.uri.replace(content._previousURI, content.uri)
			await page.save()
		})
	}
})

export default mongoose.model("Content", schema)

File: models/content.go
package models

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"golang.org/x/net/html"
)

// Content represents a content page in the system
type Content struct {
	ID                primitive.ObjectID `bson:"_id,omitempty"`
	URI               string             `bson:"uri"`
	Title             string             `bson:"title"`
	Type              string             `bson:"type"`
	Body              string             `bson:"body,omitempty"`
	Description       string             `bson:"description,omitempty"`
	Surtitle          string             `bson:"surtitle,omitempty"`
	RedirectURI       string             `bson:"redirect_uri,omitempty"`
	Hide              bool               `bson:"hide"`
	FurtherReadingURI string             `bson:"further_reading_uri,omitempty"`
	HasSublist        bool               `bson:"has_sublist"`
	Authorship        string             `bson:"authorship,omitempty"`
	Order             int                `bson:"order"`
	Contents          []ContentItem      `bson:"contents,omitempty"`
	UpdatedAt         time.Time          `bson:"updatedAt,omitempty"`
	CreatedAt         time.Time          `bson:"createdAt,omitempty"`
	
	// Virtual fields (not stored in DB)
	Breadcrumbs    []Breadcrumb       `bson:"-"`
	Directory      [][]Content        `bson:"-"`
	Selected       map[string]string  `bson:"-"`
	Depth          int                `bson:"-"`
	NextPage       *Content           `bson:"-"`
	FurtherReading *Content           `bson:"-"`
	InvalidURIs    []string           `bson:"-"`
	EditURI        string             `bson:"-"`
	Alphabetical   map[string][]Content `bson:"-"`
	Sublist        []Content          `bson:"-"`
}

// ContentItem represents a table of contents item
type ContentItem struct {
	Text string `bson:"text"`
	ID   string `bson:"id"`
}

// Breadcrumb represents a navigation breadcrumb
type Breadcrumb struct {
	Title string `bson:"title"`
	URI   string `bson:"uri"`
}

// ContentModel provides methods for content operations
type ContentModel struct {
	collection *mongo.Collection
	db         *mongo.Database
}

// NewContentModel creates a new content model
func NewContentModel(db *mongo.Database) *ContentModel {
	return &ContentModel{
		collection: db.Collection("contents"),
		db:         db,
	}
}

// FindOne finds a single content by URI
func (m *ContentModel) FindOne(ctx context.Context, uri string) (*Content, error) {
	var content Content
	err := m.collection.FindOne(ctx, bson.M{"uri": uri}).Decode(&content)
	if err != nil {
		return nil, err
	}
	return &content, nil
}

// Find finds content matching the filter
func (m *ContentModel) Find(ctx context.Context, filter bson.M, opts *options.FindOptions) (*mongo.Cursor, error) {
	return m.collection.Find(ctx, filter, opts)
}

// FindFromParentURI finds content with a specific parent URI
func (m *ContentModel) FindFromParentURI(ctx context.Context, parent string) ([]Content, error) {
	var pattern string
	if parent != "" {
		pattern = fmt.Sprintf("^%s/[^/]+$", regexp.QuoteMeta(parent))
	} else {
		pattern = "^[^/]+$"
	}
	
	filter := bson.M{"uri": bson.M{"$regex": pattern}}
	cursor, err := m.collection.Find(ctx, filter, options.Find().SetSort(bson.D{{"order", 1}, {"title", 1}, {"description", 1}, {"surtitle", 1}}))
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	
	var contents []Content
	if err = cursor.All(ctx, &contents); err != nil {
		return nil, err
	}
	return contents, nil
}

// FindFromAdjacentURI finds content adjacent to a given URI
func (m *ContentModel) FindFromAdjacentURI(ctx context.Context, uri string) ([]Content, error) {
	parent := ParentURIFragment(uri)
	return m.FindFromParentURI(ctx, parent)
}

// GetBreadcrumbs returns breadcrumbs for a content item
func (m *ContentModel) GetBreadcrumbs(ctx context.Context, content *Content) ([]Breadcrumb, error) {
	lineage := GetLineageFromURI(ParentURIFragment(content.URI))
	
	// Find all parent pages
	filter := bson.M{"uri": bson.M{"$in": lineage}}
	cursor, err := m.collection.Find(ctx, filter, options.Find().SetSort(bson.D{{"uri", 1}}))
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	
	var parents []Content
	if err = cursor.All(ctx, &parents); err != nil {
		return nil, err
	}
	
	// Convert to breadcrumbs
	breadcrumbs := make([]Breadcrumb, len(parents))
	for i, p := range parents {
		breadcrumbs[i] = Breadcrumb{
			Title: p.Title,
			URI:   p.URI,
		}
	}
	
	return breadcrumbs, nil
}

// GetChildren returns child pages of a content item
func (m *ContentModel) GetChildren(ctx context.Context, content *Content) ([]Content, error) {
	return m.FindFromParentURI(ctx, content.URI)
}

// GetNextPage finds the next page in navigation
func (m *ContentModel) GetNextPage(ctx context.Context, content *Content) (*Content, error) {
	// No next page for level 3
	if content.Type == "level3" {
		return nil, nil
	}
	
	// Build regex pattern for finding next page
	parent := ParentURIFragment(content.URI)
	pattern := fmt.Sprintf("^(%s|%s)/[^/]+$", regexp.QuoteMeta(content.URI), regexp.QuoteMeta(parent))
	
	// Determine allowed types based on current type
	var allowedTypes []string
	switch content.Type {
	case "level1":
		allowedTypes = []string{"level2", "level3"}
	case "level2":
		allowedTypes = []string{"level3"}
	default:
		allowedTypes = []string{"level1", "level2", "level3"}
	}
	
	filter := bson.M{
		"uri":   bson.M{"$regex": pattern},
		"title": content.Title,
		"type":  bson.M{"$in": allowedTypes},
	}
	
	var next Content
	err := m.collection.FindOne(ctx, filter, options.FindOne().SetSort(bson.D{{"type", 1}})).Decode(&next)
	if err == mongo.ErrNoDocuments {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	
	return &next, nil
}

// GetInvalidLinks returns invalid links in the content body
func (m *ContentModel) GetInvalidLinks(ctx context.Context, content *Content) ([]string, error) {
	links := ExtractLinksFromHTML(content.Body)
	
	// Filter out external URLs and feedback links
	var internalLinks []string
	for _, link := range links {
		if !strings.Contains(link, "://") && !strings.HasSuffix(link, "/feedback") {
			internalLinks = append(internalLinks, strings.TrimPrefix(link, "/"))
		}
	}
	
	if len(internalLinks) == 0 {
		return []string{}, nil
	}
	
	// Check which links are valid
	validLinks := make(map[string]bool)
	
	// Check content collection
	contentFilter := bson.M{"uri": bson.M{"$in": internalLinks}}
	cursor, err := m.collection.Find(ctx, contentFilter, options.Find().SetProjection(bson.M{"uri": 1}))
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	
	for cursor.Next(ctx) {
		var result struct{ URI string `bson:"uri"` }
		if err := cursor.Decode(&result); err == nil {
			validLinks["/"+result.URI] = true
		}
	}
	
	// Check videos collection
	videoColl := m.db.Collection("videos")
	videoCursor, err := videoColl.Find(ctx, contentFilter, options.Find().SetProjection(bson.M{"uri": 1}))
	if err == nil {
		defer videoCursor.Close(ctx)
		for videoCursor.Next(ctx) {
			var result struct{ URI string `bson:"uri"` }
			if err := videoCursor.Decode(&result); err == nil {
				validLinks["/"+result.URI] = true
			}
		}
	}
	
	// Check images collection
	imageColl := m.db.Collection("images")
	imageCursor, err := imageColl.Find(ctx, contentFilter, options.Find().SetProjection(bson.M{"uri": 1}))
	if err == nil {
		defer imageCursor.Close(ctx)
		for imageCursor.Next(ctx) {
			var result struct{ URI string `bson:"uri"` }
			if err := imageCursor.Decode(&result); err == nil {
				validLinks["/"+result.URI] = true
			}
		}
	}
	
	// Find invalid links
	var invalidLinks []string
	for _, link := range internalLinks {
		if !validLinks["/"+link] && !validLinks[link] {
			if !strings.HasPrefix(link, "/") {
				link = "/" + link
			}
			invalidLinks = append(invalidLinks, link)
		}
	}
	
	return invalidLinks, nil
}

// GetMatchedParagraph finds a paragraph matching the given pattern
func (m *ContentModel) GetMatchedParagraph(content *Content, pattern *regexp.Regexp) []string {
	doc, err := html.Parse(strings.NewReader(content.Body))
	if err != nil {
		return nil
	}
	
	var search func(*html.Node) []string
	search = func(n *html.Node) []string {
		if n.Type == html.ElementNode {
			switch n.Data {
			case "p", "li", "td", "h1", "h2", "h3", "h4", "h5":
				text := extractText(n)
				if matches := pattern.FindStringSubmatch(text); matches != nil {
					return matches
				}
			}
		}
		
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			if result := search(c); result != nil {
				return result
			}
		}
		return nil
	}
	
	return search(doc)
}

// Helper functions

// ParentURIFragment returns the parent URI of a given URI
func ParentURIFragment(uri string) string {
	parts := strings.Split(uri, "/")
	if len(parts) <= 1 {
		return ""
	}
	return strings.Join(parts[:len(parts)-1], "/")
}

// GetLineageFromURI returns all parent URIs up to root
func GetLineageFromURI(uri string) []string {
	var fragments []string
	parent := uri
	for parent != "" {
		fragments = append(fragments, parent)
		parent = ParentURIFragment(parent)
	}
	
	// Reverse the array
	for i, j := 0, len(fragments)-1; i < j; i, j = i+1, j-1 {
		fragments[i], fragments[j] = fragments[j], fragments[i]
	}
	
	return fragments
}

// NormalizeURI normalizes a URI to acceptable format
func NormalizeURI(uri string) string {
	// All lowercase
	uri = strings.ToLower(uri)
	// Convert spaces and underscores to dashes
	uri = regexp.MustCompile(`[_ -]+`).ReplaceAllString(uri, "-")
	// Remove duplicate slashes
	uri = regexp.MustCompile(`/+`).ReplaceAllString(uri, "/")
	// Remove leading/trailing slashes or dashes
	uri = regexp.MustCompile(`^[/-]+|[/-]+$`).ReplaceAllString(uri, "")
	// Remove invalid characters
	uri = regexp.MustCompile(`[^a-z0-9-/]+`).ReplaceAllString(uri, "")
	
	return uri
}

// ExtractLinksFromHTML extracts all links from HTML content
func ExtractLinksFromHTML(htmlContent string) []string {
	doc, err := html.Parse(strings.NewReader(htmlContent))
	if err != nil {
		return []string{}
	}
	
	var links []string
	var extract func(*html.Node)
	extract = func(n *html.Node) {
		if n.Type == html.ElementNode && n.Data == "a" {
			for _, attr := range n.Attr {
				if attr.Key == "href" {
					links = append(links, attr.Val)
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			extract(c)
		}
	}
	
	extract(doc)
	return links
}

// extractText recursively extracts text from an HTML node
func extractText(n *html.Node) string {
	if n.Type == html.TextNode {
		return n.Data
	}
	
	var text string
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		text += extractText(c)
	}
	return text
}